File: Source/GITRepo.m
//
//  GITRepo.m
//  Git.framework
//
//  Created by Geoff Garside on 14/09/2009.
//  Copyright 2009 Geoff Garside. All rights reserved.
//

#import "GITRepo.h"
#import "GITError.h"
#import "GITRefResolver.h"
#import "GITBranch.h"
#import "GITObject.h"
#import "GITCommit.h"
#import "GITObjectHash.h"
#import "GITPackObject.h"
#import "GITLooseObject.h"
#import "GITPackCollection.h"
#import "GITCommitEnumerator.h"
#import "GITGraph.h"
#import "GITGraphNode.h"
#import "GITRevList.h"


@interface GITRepo ()
@property (copy) NSString *objectsDirectory;
@property (retain) GITPackCollection *packCollection;

- (BOOL)rootExists;
- (BOOL)rootIsAccessible;
- (BOOL)rootDoesLookSane;

- (NSData *)dataForHeadFile;
- (NSData *)dataForConfigFile;
- (NSData *)dataForDescriptionFile;
- (NSData *)dataForInfoExcludeFile;
- (NSArray *)arrayOfSkeletonDirectories;
- (BOOL)createDirectorySkeletonAtPath: (NSString *)theRoot error: (NSError **)theError;

@end

@implementation GITRepo

@synthesize root;
@synthesize bare;
@synthesize refResolver;
@synthesize packCollection;
@synthesize objectsDirectory;

+ (GITRepo *)repo {
    return [[[GITRepo alloc] initWithRoot:[[NSFileManager defaultManager] currentDirectoryPath] error: NULL] autorelease];
}
+ (GITRepo *)repoWithRoot: (NSString *)theRoot {
    return [[[GITRepo alloc] initWithRoot: theRoot error: NULL] autorelease];
}
+ (GITRepo *)repoWithRoot: (NSString *)theRoot error: (NSError **)theError {
    return [[[GITRepo alloc] initWithRoot: theRoot error: theError] autorelease];
}
+ (GITRepo *)createRepoAtPath: (NSString *)path {
    return [[[GITRepo alloc] initAtPath: path error: NULL] autorelease];
}
+ (GITRepo *)createRepoAtPath: (NSString *)path error: (NSError **)theError {
    return [[[GITRepo alloc] initAtPath: path error: theError] autorelease];
}

- (id)initAtPath: (NSString *)path {
    return [self initAtPath: path error: NULL];
}
- (id)initAtPath: (NSString *)path error: (NSError **)theError {
    if ( ![self createDirectorySkeletonAtPath:path error:theError] )
        return nil;
    return [self initWithRoot:path error:theError];
}

- (id)initWithRoot: (NSString *)theRoot {
    return [self initWithRoot: theRoot error: NULL];
}
- (id)initWithRoot: (NSString *)theRoot error: (NSError **)theError {
    if ( ![super init] )
        return nil;
    
    self.root = [theRoot stringByStandardizingPath];

    if ( !(self.bare = [self.root hasSuffix:@".git"]) ) {
        self.root = [self.root stringByAppendingPathComponent:@".git"];
    }

    if ( ![self rootExists] ) {
        GITError(theError, GITRepoErrorRootDoesNotExist, NSLocalizedString(@"Path to repository does not exist", @"GITRepoErrorRootDoesNotExist"));
        [self release];
        return nil;
    }
    if ( ![self rootIsAccessible] ) {
        GITError(theError, GITRepoErrorRootNotAccessible, NSLocalizedString(@"Path to repository could not be opened, check permissions", @"GITRepoErrorRootNotAccessible"));
        [self release];
        return nil;
    }
    if ( ![self rootDoesLookSane] ) {
        GITError(theError, GITRepoErrorRootInsane, NSLocalizedString(@"Path does not appear to be a git repository", @"GITRepoErrorRootInsane"));
        [self release];
        return nil;
    }

    self.objectsDirectory = [self.root stringByAppendingPathComponent:@"objects"];
    self.packCollection = [GITPackCollection collectionWithContentsOfDirectory:[self.objectsDirectory stringByAppendingPathComponent:@"pack"] error:theError];
    if ( !packCollection ) {
        [self release];
        return nil;
    }
    
    return self;
}

- (void)dealloc {
    self.root = nil;
    self.objectsDirectory = nil;
    self.packCollection = nil;
    self.refResolver = nil;
    [super dealloc];
}

- (BOOL)rootExists {
    BOOL isDirectory;
    return [[NSFileManager defaultManager] fileExistsAtPath:self.root isDirectory:&isDirectory] && isDirectory;
}

- (BOOL)rootIsAccessible {
    return [[NSFileManager defaultManager] isReadableFileAtPath:self.root] &&
        [[NSFileManager defaultManager] isWritableFileAtPath:self.root];
}

- (BOOL)rootDoesLookSane {
    NSString *path;
    BOOL isSane = NO;
    BOOL isDirectory;

    NSFileManager *fm = [NSFileManager defaultManager];
    NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];

    NSArray *fileChecks = [NSArray arrayWithObjects: @"HEAD", @"config", @"index", nil];
    for ( NSString *pathComponent in fileChecks ) {
        isDirectory = NO;
        path = [self.root stringByAppendingPathComponent: pathComponent];
        if ( ![fm fileExistsAtPath: path isDirectory:&isDirectory] || isDirectory )
            goto done;
    }

    NSArray *dirChecks  = [NSArray arrayWithObjects: @"refs", @"objects", nil];
    for ( NSString *pathComponent in dirChecks ) {
        isDirectory = NO;
        path = [self.root stringByAppendingPathComponent: pathComponent];
        if ( ![fm fileExistsAtPath: path isDirectory:&isDirectory] || !isDirectory )
            goto done;
    }

    isSane = YES;

done:
    [pool drain];
    return isSane;
}

- (GITRefResolver *)refResolver {
    if ( !refResolver )
        self.refResolver = [GITRefResolver resolverForRepo:self];
    return refResolver;
}

- (NSArray *)branches {
    NSArray *headRefs = [[self refResolver] headRefs];
    NSMutableArray *branches = [NSMutableArray arrayWithCapacity:[headRefs count]];

    for ( GITRef *ref in headRefs ) {
        [branches addObject:[GITBranch branchFromRef:ref]];
    }

    return [[branches copy] autorelease];
}

- (NSArray *)remoteBranches {
    NSArray *remoteRefs = [[self refResolver] remoteRefs];
    NSMutableArray *branches = [NSMutableArray arrayWithCapacity:[remoteRefs count]];

    for ( GITRef *ref in remoteRefs ) {
        [branches addObject:[GITBranch branchFromRef:ref]];
    }

    return [[branches copy] autorelease];
}

- (NSArray *)tags {
    return [[self refResolver] tagRefs];
}

- (GITObject *)objectWithSha1: (GITObjectHash *)objectHash error: (NSError **)error {
    // Need to load it from the file system
    GITLooseObject *looseObject = [GITLooseObject looseObjectWithSha1:objectHash from:self.objectsDirectory error:error];
    if ( looseObject )      // Should really return the error if it was something bad
        return [looseObject objectInRepo:self error:error];

    // Need to load it from the pack collection
    GITPackObject *packObject = [self.packCollection unpackObjectWithSha1:objectHash error:error];
    if ( !packObject )
        return nil;
    return [packObject objectInRepo:self error:error];
}

- (GITCommit *)head {
    return (GITCommit *)[[[self refResolver] resolveRefWithName:@"HEAD"] target];
}

- (GITCommitEnumerator *)enumerator {
    return [GITCommitEnumerator enumeratorFromCommit:[self head]];
}

- (GITCommitEnumerator *)enumeratorWithMode: (GITCommitEnumeratorMode)mode {
    return [GITCommitEnumerator enumeratorFromCommit:[self head] mode:mode];
}

- (GITRevList *)revList {
    return [self revListFromCommit:[self head]];
}
- (GITRevList *)revListFromCommit: (GITCommit *)head {
    return [GITRevList revListWithCommit:head];
}

- (NSArray *)revListSortedByDate {
    return [[self revList] arrayOfCommitsSortedByDate];
}
- (NSArray *)revListSortedByTopology {
    return [[self revList] arrayOfCommitsSortedByTopology];
}
- (NSArray *)revListSortedByTopologyAndDate {
    return [[self revList] arrayOfCommitsSortedByTopologyAndDate];
}

#pragma mark Creating Skeleton
- (NSData *)dataForHeadFile {
    return [[NSString stringWithString:@"ref: refs/heads/master\n"] dataUsingEncoding:NSASCIIStringEncoding];
}
- (NSData *)dataForConfigFile {
    NSString *str = [NSString stringWithString:@"[core]\n  repositoryformatversion = 0\n  filemode = true\n  bare = false\n  logallrefupdates = true\n  ignorecase = true\n"];
    return [str dataUsingEncoding:NSASCIIStringEncoding];
}
- (NSData *)dataForDescriptionFile {
    return [[NSString stringWithString:@"Unnamed repository; edit this file 'description' to name the repository.\n"] dataUsingEncoding:NSASCIIStringEncoding];
}
- (NSData *)dataForInfoExcludeFile {
    NSString *str = [NSString stringWithString:@"# git ls-files --others --exclude-from=.git/info/exclude\n# Lines that start with '#' are comments.\n# For a project mostly in C, the following would be a good set of\n# exclude patterns (uncomment them if you want to use them):\n# *.[oa]\n# *~\n"];
    return [str dataUsingEncoding:NSASCIIStringEncoding];

}
- (NSArray *)arrayOfSkeletonDirectories {
    return [NSArray arrayWithObjects:@"branches", @"hooks", @"info", @"objects/info", @"objects/pack", @"refs/heads", @"refs/tags", nil];
}
- (BOOL)createDirectorySkeletonAtPath: (NSString *)path error: (NSError **)theError {
    NSFileManager *fm = [[NSFileManager alloc] init];

    NSString *gitRoot;
    if ( [path hasSuffix:@".git"] ) {
        gitRoot = [path copy];
    } else {
        gitRoot = [path stringByAppendingPathComponent:@".git"];
    }

    if ( [fm fileExistsAtPath:gitRoot] ) {
        GITError(theError, GITRepoErrorSkeletonExists, NSLocalizedString(@"Directory for skeleton exists", @"GITRepoErrorSkeletonExists"));
        return NO;
    }

    NSError *directoryError;
    for ( NSString *dir in [self arrayOfSkeletonDirectories] ) {
        if ( ![fm createDirectoryAtPath:[gitRoot stringByAppendingPathComponent:dir]
                withIntermediateDirectories:YES attributes:nil error:&directoryError] ) {
            GITErrorWithInfo(theError, GITRepoErrorSkeletonCreationFailed, NSUnderlyingErrorKey, directoryError,
                             NSLocalizedDescriptionKey, NSLocalizedString(@"Failed to create skeleton directory", @"GITRepoErrorCreationFailed directory"), nil);
            goto cleanup;
        }
    }

    if ( ![fm createFileAtPath:[gitRoot stringByAppendingPathComponent:@"HEAD"]
                      contents:[self dataForHeadFile] attributes:nil] ) {
        GITError(theError, GITRepoErrorSkeletonCreationFailed, NSLocalizedString(@"Failed to create HEAD file", @"GITRepoErrorCreationFailed HEAD file"));
        goto cleanup;
    }

    if ( ![fm createFileAtPath:[gitRoot stringByAppendingPathComponent:@"config"]
                      contents:[self dataForConfigFile] attributes:nil] ) {
        GITError(theError, GITRepoErrorSkeletonCreationFailed, NSLocalizedString(@"Failed to create config file", @"GITRepoErrorCreationFailed config file"));
        goto cleanup;
    }

    if ( ![fm createFileAtPath:[gitRoot stringByAppendingPathComponent:@"description"]
                      contents:[self dataForDescriptionFile] attributes:nil] ) {
        GITError(theError, GITRepoErrorSkeletonCreationFailed, NSLocalizedString(@"Failed to create description file", @"GITRepoErrorCreationFailed description file"));
        goto cleanup;
    }

    if ( ![fm createFileAtPath:[gitRoot stringByAppendingPathComponent:@"info/exclude"]
                      contents:[self dataForInfoExcludeFile] attributes:nil] ) {
        GITError(theError, GITRepoErrorSkeletonCreationFailed, NSLocalizedString(@"Failed to create info/exclude file", @"GITRepoErrorCreationFailed info/exclude file"));
        goto cleanup;
    }

    [fm release];
    return YES;

cleanup:
    [fm removeItemAtPath:gitRoot error:NULL];
    [fm release];
    return NO;
}

@end

File: Test/Unit/GITRepo.rb
describe "GITRepo +repo" do
  before do
    Dir.chdir(simple_repository.root) do
      @repo = GITRepo.repo
    end
  end

  should "open the repo in the current directory" do
    @repo.root.should.equal "#{simple_repository.root}/.git"
  end
end

describe "GITRepo +repoWithRoot:" do
  describe "with valid repository" do
    before do
      @repo = GITRepo.repoWithRoot(simple_repository.root)
    end

    should "open repo in correct directory" do
      @repo.root.should.equal "#{simple_repository.root}/.git"
    end
  end
  describe "with nonexistent repository" do
    before do
      @repo = GITRepo.repoWithRoot("/nonexistent")
    end

    should "be nil" do
      @repo.should.be.nil
    end
  end
end

describe "GITRepo +repoWithRoot:error:" do
  before do
    @error = Pointer.new(:object)
  end
  describe "with valid repository" do
    before do
      @repo = GITRepo.repoWithRoot(simple_repository.root, error:@error)
    end

    should "open repo at " do
      @repo.root.should.equal "#{simple_repository.root}/.git"
    end
    should "not raise an error" do
      @error[0].should.be.nil
    end
  end
  describe "with nonexistent path" do
    before do
      @repo = GITRepo.repoWithRoot('/nonexistent', error:@error)
    end

    should "return nil" do
      @repo.should.be.nil
    end
    should "raise an error" do
      @error[0].should.not.be.nil
    end
    should "raise a path error" do
      @error[0].code.should.equal GITRepoErrorRootDoesNotExist
    end
  end
  describe "with no permission to path" do
    before do
      `chmod a-w #{simple_repository.root}/.git`
      @repo = GITRepo.repoWithRoot(simple_repository.root, error:@error)
    end

    after do
      # So that we can remove the repository
      `chmod a+w #{simple_repository.root}/.git`
    end

    should "return nil" do
      @repo.should.be.nil
    end
    should "raise an error" do
      @error[0].should.not.be.nil
    end
    should "raise a path error" do
      @error[0].code.should.equal GITRepoErrorRootNotAccessible
    end
  end
  describe "with insane looking repository" do
    before do
      FileUtils.rm_rf("#{simple_repository.root}/.git/refs")
      @repo = GITRepo.repoWithRoot(simple_repository.root, error:@error)
    end

    should "return nil" do
      @repo.should.be.nil
    end
    should "raise an error" do
      @error[0].should.not.be.nil
    end
    should "raise a sanity error" do
      @error[0].code.should.equal GITRepoErrorRootInsane
    end
  end
  describe "with missing .git/branches directory" do
    before do
      FileUtils.rm_rf("#{simple_repository.root}/.git/branches")
      @repo = GITRepo.repoWithRoot(simple_repository.root, error:@error)
    end

    should "open repo" do
      @repo.root.should.equal "#{simple_repository.root}/.git"
    end
    should "not raise an error" do
      @error[0].should.be.nil
    end
  end
end

describe "GITRepo -branches" do
  before do
    @repo = simple_repository.git_repo
    @branches = @repo.branches.map(&:name)
  end

  should "include 'master'" do
    @branches.should.include('master')
  end
  should "include 'another'" do
    @branches.should.include('another')
  end
end

describe "GITRepo -objectWithSha1:error:" do
  before do
    @err = Pointer.new(:object)
    @repo = simple_repository.git_repo
    @sha1 = GITObjectHash.objectHashWithString(simple_repository.commit('Update testfile.txt').sha)
    @commit = @repo.objectWithSha1(@sha1, error:@err)
  end

  should "not be nil" do
    @commit.should.not.be.nil
  end
  should "not have an error" do
    @err[0].should.be.nil
  end
  should "have sha1" do
    @commit.sha1.unpackedString.should == @sha1.unpackedString
  end
end

describe "GITRepo Enumerators" do
  before do
    @repo = simple_repository.git_repo
  end
  describe "-enumerator" do
    before do
      @enum = @repo.enumerator
    end
    should "not be nil" do
      @enum.should.not.be.nil
    end
  end
  describe "-enumeratorWithMode:" do
    before do
      @enum = @repo.enumeratorWithMode(GITCommitEnumeratorBreadthFirstMode)
    end
    should "not be nil" do
      @enum.should.not.be.nil
    end
  end
end

describe "GITRepo -head" do
  before do
    @repo = simple_repository.git_repo
  end
  should "not be nil" do
    @repo.head.should.not.be.nil
  end
end

describe "GITRepo -tags" do
  before do
    @repo = simple_repository.git_repo
    @tags = @repo.tags
  end
  should 'not be empty' do
    @tags.count.should.be > 0
  end
  should 'contain GITTag objects' do
    @tags.each do |tag|
      tag.should.be.kind_of(GITTag)
    end
  end
end

describe "GITRepo -revList" do
  before do
    @repo = graph_repository.git_repo
    @revList = @repo.revList
  end
  should "not be nil" do
    @revList.should.not.be.nil
  end
  should "be a GITRevList object" do
    @revList.should.be.kind_of GITRevList
  end
end

describe "GITRepo -revListFromCommit:" do
  before do
    @repo = graph_repository.git_repo
    @revList = @repo.revListFromCommit(@repo.head)
  end
  should "not be nil" do
    @revList.should.not.be.nil
  end
  should "be a GITRevList object" do
    @revList.should.be.kind_of GITRevList
  end
end

describe "GITRepo Rev-List Methods" do
  before do
    @repo = graph_repository.git_repo
  end
  describe "-revListSortedByDate" do
    before do
      @list = @repo.revListSortedByDate
      @expected = graph_repository.git("rev-list master").split("\n").map { |l| l.strip }
    end
    should "return expected ordering of commits" do
      @list.map { |c| c.sha1.unpackedString }.should == @expected
    end
  end
  describe "-revListSortedByTopology" do
    before do
      @list = @repo.revListSortedByTopology
      @expected = graph_repository.git("rev-list --topo-order master").split("\n").map { |l| l.strip }
    end
    should "return expected ordering of commits" do
      @list.map { |c| c.sha1.unpackedString }.should == @expected
    end
  end
  describe "-revListSortedByTopologyAndDate" do
    before do
      @list = @repo.revListSortedByTopologyAndDate
      @expected = graph_repository.git("rev-list --date-order master").split("\n").map { |l| l.strip }
    end
    should "return expected ordering of commits" do
      @list.map { |c| c.sha1.unpackedString }.should == @expected
    end
  end
end

describe "GITRepo +createRepoAtPath:error:" do
  before do
    @err  = Pointer.new(:object)
    @path = File.expand_path("Created", tmp_dir)
    @root = File.expand_path(".git", @path)
    @repo = GITRepo.createRepoAtPath(@path, error:@err)
  end
  after do
    FileUtils.rm_r(@path)
  end
  should "not raise an error" do
    unless @err[0].nil?
      puts @err[0].localizedDescription
    end
    @err[0].should.be.nil
  end
  should "not return nil" do
    @repo.should.not.be.nil
  end
  should "create .git directory" do
    File.exists?(@root).should.be.true
  end
  should "create skeleton directories and files" do
    %w(branches config description HEAD hooks info info/exclude objects objects/info objects/pack refs refs/heads refs/tags).each do |f|
      File.exists?(File.join(@root, f)).should.be.true
    end
  end
end
